### ISSUE TITLE
Read non standard fields in camtraptor

### ISSUE THREAD

**Issue by damianooldoni** [inbo/camtraptor#185]
I got this question from @MarcusRowcliffe via email: 

> One thing about this that I find odd – there are three crucial new fields in the observations data that we need for analysis – speed, distance and angle. These are named properly in the observations.csv file that comes with the zip, but in the observations table in the package list they are named with default values – X22 etc. Is this something you can fix? I understand these are not currently recognised camtrap-dp fields, so maybe that is the problem?

**Comment by damianooldoni** [comment]
By opening the datapackage provided, I see in the `datapackage.json` the following:

```json
    {
      "name": "observations",
      "path": "observations.csv",
      "profile": "tabular-data-resource",
      "format": "csv",
      "mediatype": "text/csv",
      "encoding": "utf-8",
      "schema": "https://raw.githubusercontent.com/tdwg/camtrap-dp/0.1.6/observations-table-schema.json"
    }
```

To read the three new fields, my suggestion is to modify the schema by providing another URL with the adapted schema or just copy paste it. The `read_camtrapdp_dp()` doesn't need any change, I think..
I will test it now and post the result.

**Comment by peterdesmet** [comment]
Yes, that is a reasonable approach until these fields are included in Camtrap DP.

**Comment by damianooldoni** [comment]
I have just checked it and my solution works 🥳 

Please, replace your datapackage.json file with the csv in attachment (no json extension supported for GitHub attachments) Notice you can leave the extension as it is (`.csv`) or set it back to `.json` if you prefer. 

Then, read the datapackage as you are used to via `read_camtrap_dp()`.

Here my attempt:

```r
> library(camtraptor)
> test <- read_camtrap_dp(file = "../../../Downloads/camera-model-calibration-20221114172535/datapackage.json")
Please make sure you have the right to access data from this Data Package for your intended use.
Follow applicable norms or requirements to credit the dataset and its authors.
> obs <- test$data$observations                                                                                                 
> names(obs)
 [1] "observationID"            "deploymentID"             "sequenceID"              
 [4] "mediaID"                  "timestamp"                "observationType"         
 [7] "cameraSetup"              "taxonID"                  "taxonIDReference"        
[10] "scientificName"           "taxonRank"                "vernacularNames.en"      
[13] "vernacularNames.nl"       "count"                    "countNew"                
[16] "lifeStage"                "sex"                      "behaviour"               
[19] "individualID"             "classificationMethod"     "classifiedBy"            
[22] "classificationTimestamp"  "classificationConfidence" "comments"                
[25] "_id"                      "speed"                    "distance"                
[28] "angle" 
```

As you can see, the last three columns are read correctly. Notice that in my datapackage.json file I attempted to fill the **description** argument of those three fields as follows:

```
            {
            "name": "speed",
            "type": "number",
            "format": "default",
            "description": "Speed of the detected animal in m/s",
            "example": "12.5",
            "constraints": {
              "required": false
            }
          },
          {
            "name": "distance",
            "type": "number",
            "format": "default",
            "description": "Distance from the camera of the detected animal in meters",
            "example": "5.2",
            "constraints": {
              "required": false
            }
          },
          {
            "name": "angle",
            "type": "number",
            "format": "default",
            "description": "Angle of the detected animal from camera",
            "example": "5.2",
            "constraints": {
              "required": false
            }
          }
``` 

You can of course improve it as I have no idea whether you are using meters and I don't know which unit you use for angles.
You can also add better constraints (for angles for example, if you use degrees: min and max -90, +90) but it's just a bonus.

I hope this helps! Let me know.

[datapackage.csv](https://github.com/inbo/camtraptor/files/10246645/datapackage.csv)

**Comment by MarcusRowcliffe** [comment]
Thanks Damiano. I have a couple of problems with this solution - although not sure if I have misunderstood something. 
First, as discussed, the distance from camera should be named radius, not distance.
Second, as I understand it, the modified datapackage file contains all the original project-specific information (locations, contributors, species etc), so we couldn't simply pass this file to anyone for use with another projects. Am I missing something?

**Comment by damianooldoni** [comment]
Thanks @MarcusRowcliffe! Of course, I understand the point. I give you here below a step-by-step solution.

1. Rename `distance` to `radius` manually in the `csv` **before** importing the datapackage
2. I saved the improved table schema for observations (using radius instead of distance!) in a public gist called [observations-table-schema_improved.json](https://gist.githubusercontent.com/damianooldoni/c50e886f9a7ff4e4ab4e551d635b77ae/raw/c54d685dd14426d79ceebc6cae80363820b92e48/observations-table-schema_improved.json). A gist is a place to drop "stuff" via GitHub. As repos, they can be public or private. Use please the URL of the gist instead of the original `"https://raw.githubusercontent.com/tdwg/camtrap-dp/0.1.6/observations-table-schema.json"` in your `datapackage.json`. The last part of your json file should look like this:

```json
{
      "name": "observations",
      "path": "observations.csv",
      "profile": "tabular-data-resource",
      "format": "csv",
      "mediatype": "text/csv",
      "encoding": "utf-8",
      "schema": "https://gist.githubusercontent.com/damianooldoni/c50e886f9a7ff4e4ab4e551d635b77ae/raw/c54d685dd14426d79ceebc6cae80363820b92e48/observations-table-schema_improved.json"
    }
```

I hope this works. Please, let me know!

**Comment by peterdesmet** [comment]
It is important to note that this solution is a **temporary hack** until a new version of Camtrap DP comes out that has the 3 desired fields.

**Comment by MarcusRowcliffe** [comment]
Many thanks Damiano. That works! But I would rather not have to walk all the diverse people currently using these tools through the process. My current temporary hack (a wrapper function for read_camtrap_dp that renames the columns within R once imported) seems to be working and is more seamless for the user, so will stick with that for now I think. Given that, I don't think there is a hurry, but roughly how long do you think it will take to get these new fields incorporated as standard?

**Comment by damianooldoni** [comment]
@peterdesmet, I think you can answer better than me to this question. Actually it's important for me to know as well as my "hack" solution at data package level is indeed not the handiest on the long term.

So, question of weeks: I will not change anything in `read_camtrap_dp()`. Hacking the datapackage metadata as I have shown is still my solution to this issue (users can choose to rename the columns in their own wrapper function as @MarcusRowcliffe did).

Question of months: I will "hack" `read_camtrap_dp()` as asked by @MarcusRowcliffe. 
Thanks.

**Comment by peterdesmet** [comment]
It is difficult to assess when a new version of the standard will be out. It always takes longer than expected and Agouti needs to then be updated as well.

I would therefore suggest to change `camtraptor::read_camtrap_dp()` now, so users a) don't need to fiddle with the `datapackage.json` to link to another table schema and b) don't require a `read_camtrap_dp2()` function.

I would implement it fairly similar to how @MarcusRowcliffe has done it in his `read_camtrap_dp2()` function.

The only part that this approach does not allow is validation of a Camtrap DP with frictionless-py, but most users don't do that anyway.

**Comment by damianooldoni** [comment]
Thanks @peterdesmet. In this case I definitely add a patch to `read_camtrap_dp()` function. I plan it for tomorrow 👍

**Comment by MarcusRowcliffe** [comment]
Many thanks @damianooldoni

**Comment by damianooldoni** [comment]
Solved via #186. @MarcusRowcliffe, @LynnPallemaerts, @jimcasaer and all other users of Agouti exports: you can read the data pacakges with the extra non standard fields using camtraptor's function `read_camtrap_dp()`. A message (no warning!) is printed on console notifying you about the presence of the three new fields.
